libcxx/include/__flat_map/flat_map.h

```diff
@@ -0,0 +1,1280 @@
+    return __binary_search_emplace_impl(std::move(__pair));
+  }
+
+  template <class... _Args>
+    requires is_constructible_v<pair<key_type, mapped_type>, _Args...>
+  _LIBCPP_HIDE_FROM_ABI iterator emplace_hint(const_iterator __hint, _Args&&... __args) {
+    std::pair<key_type, mapped_type> __pair(std::forward<_Args>(__args)...);
+    if (__is_hint_correct(__hint, __pair.first)) {
+      if (__compare_(__pair.first, __hint->first)) {
+        return __emplace_impl(__hint, std::move(__pair));
+      } else {
+        // key equals
+        auto __dist = __hint - cbegin();
+        return iterator(__containers_.keys.begin() + __dist, __containers_.values.begin() + __dist);
+      }
+    } else {
+      return __binary_search_emplace_impl(std::move(__pair)).first;
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI pair<iterator, bool> insert(const value_type& __x) { return emplace(__x); }
+
+  _LIBCPP_HIDE_FROM_ABI pair<iterator, bool> insert(value_type&& __x) { return emplace(std::move(__x)); }
+
+  _LIBCPP_HIDE_FROM_ABI iterator insert(const_iterator __hint, const value_type& __x) {
+    return emplace_hint(__hint, __x);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI iterator insert(const_iterator __hint, value_type&& __x) {
+    return emplace_hint(__hint, std::move(__x));
+  }
+
+  template <class _Pp>
+    requires is_constructible_v<pair<key_type, mapped_type>, _Pp>
+  _LIBCPP_HIDE_FROM_ABI pair<iterator, bool> insert(_Pp&& __x) {
+    return emplace(std::forward<_Pp>(__x));
+  }
+
+  template <class _Pp>
+    requires is_constructible_v<pair<key_type, mapped_type>, _Pp>
+  _LIBCPP_HIDE_FROM_ABI iterator insert(const_iterator __hint, _Pp&& __x) {
+    return emplace_hint(__hint, std::forward<_Pp>(__x));
+  }
+
+  template <input_iterator _InputIterator>
+  _LIBCPP_HIDE_FROM_ABI void insert(_InputIterator __first, _InputIterator __last) {
+    if constexpr (sized_sentinel_for<_InputIterator, _InputIterator>) {
+      __reserve_impl(__last - __first);
+    }
+
+    for (; __first != __last; ++__first) {
+      __binary_search_emplace_impl(value_type(*__first));
+    }
+  }
+
+  template <input_iterator _InputIterator>
+  void insert(sorted_unique_t, _InputIterator __first, _InputIterator __last) {
+    if constexpr (sized_sentinel_for<_InputIterator, _InputIterator>) {
+      __reserve_impl(__last - __first);
+    }
+
+    auto __it = begin();
+    while (__first != __last) {
+      value_type __pair(*__first);
+      auto __end = end();
+      __it       = ranges::lower_bound(__it, __end, __pair.first, __compare_, [](const auto& __p) -> decltype(auto) {
+        return std::get<0>(__p);
+      });
+      if (__it == __end || __compare_(__pair.first, __it->first)) {
+        __it = __emplace_impl(__it, std::move(__pair));
+      }
+      ++__it;
+      ++__first;
+    }
+  }
+
+  template <_ContainerCompatibleRange<value_type> _Range>
+  _LIBCPP_HIDE_FROM_ABI void insert_range(_Range&& __range) {
```

Comment:
I would be consistent with `insert` here and not reserve. It does feel like an overload with `sorted_unique` is missing though. Let's try to file a LWG issue for this one too and see if they ask for a paper. We'd want to DR that anyway.